solr/core/src/test/org/apache/solr/handler/admin/api/CreateCoreApiTest.java

```diff
@@ -0,0 +1,153 @@
+    Path subHome = instDir.resolve("conf");
+    Files.createDirectories(subHome);
+    String srcDir = SolrTestCaseJ4.TEST_HOME() + "/collection1/conf";
+    Files.copy(Path.of(srcDir, "schema-tiny.xml"), subHome.resolve("schema_ren.xml"));
+    Files.copy(Path.of(srcDir, "solrconfig-minimal.xml"), subHome.resolve("solrconfig_ren.xml"));
+    Files.copy(
+        Path.of(srcDir, "solrconfig.snippet.randomindexconfig.xml"),
+        subHome.resolve("solrconfig.snippet.randomindexconfig.xml"));
+    final CoreContainer cores = h.getCoreContainer();
+    cores.getAllowPaths().add(workDir);
+    // cores.getAllowPaths().add(workDir);
+    coreAdminHandler = new CoreAdminHandler(cores);
+    SolrQueryResponse response = new SolrQueryResponse();
+    createCore =
+        new CreateCore(
+            coreAdminHandler.getCoreContainer(),
+            coreAdminHandler.getCoreAdminAsyncTracker(),
+            req(),
+            response);
+  }
+
+  @Test
+  public void test_createCore_with_existent_configSet() {
+    CreateCoreRequestBody createCoreRequestBody = new CreateCoreRequestBody();
+    createCoreRequestBody.name = coreName;
+    createCoreRequestBody.configSet = "_default";
+    createCoreRequestBody.dataDir = "data_demo";
+    createCoreRequestBody.instanceDir = instDir.toAbsolutePath().toString();
+    createCore.createCore(createCoreRequestBody, CREATE_CORE_NAME);
+    Path dataDir = instDir.resolve("data_demo");
+    assertTrue(Files.exists(dataDir));
+  }
+
+  @Test
+  public void testReportError_createCore_non_existent_configSet() {
+    CreateCoreRequestBody createCoreRequestBody = new CreateCoreRequestBody();
+    createCoreRequestBody.name = coreName;
+    createCoreRequestBody.configSet = "_default_non_existent_configSet";
+    createCoreRequestBody.instanceDir = instDir.toAbsolutePath().toString();
+    assertThrows(
+        "Could not load configuration from directory",
+        SolrException.class,
+        () -> createCore.createCore(createCoreRequestBody, CREATE_CORE_NAME));
+  }
+
+  @Test
+  public void testReportError_two_thread_simultaneously_createCore() throws InterruptedException {
+    final CreateCoreRequestBody createCoreRequestBody = new CreateCoreRequestBody();
+    createCoreRequestBody.name = coreName;
+    createCoreRequestBody.instanceDir = instDir.toAbsolutePath().toString();
+    createCoreRequestBody.config = "solrconfig_ren.xml";
+    createCoreRequestBody.schema = "schema_ren.xml";
+    createCoreRequestBody.dataDir = "data_demo";
+    final AtomicReference<Exception> exp = new AtomicReference<>();
+    final Runnable coreCreationCmd =
+        () -> {
+          try {
+            createCore.createCore(createCoreRequestBody, CREATE_CORE_NAME);
+          } catch (Exception e) {
+            exp.set(e);
+          }
+        };
+
+    Thread firstCoreCreationThread = new Thread(coreCreationCmd);
+    Thread secCoreCreationThread = new Thread(coreCreationCmd);
+    firstCoreCreationThread.start();
```

Comment:
[0] Might simplify this code slightly to create a ExecutorService via `ExecutorUtil`, rather than directly managing threads?  Just a suggestion...

Also you're likely aware of this, but this setup doesn't guarantee that the core creations happen "simultaneously".  "First" might be finished by the time that "second" is started.  It's still a good test to have IMO, just might need a different name.